packages/babel-plugin-minify-constant-folding/__tests__/fixtures/member-access/actual.js

```diff
@@ -0,0 +1,13 @@
+({foo: 'oof', bar: 'rab'}).foo;
+({bar: 'rab'}).foo;
+({['foo']: 'oof', bar: 'rab'}).foo;
+({['foo']: 'oof', bar: 'rab'}).baz;
+({bar: 'rab'}).toString;
+({
+  foo: 'bar',
+  [Math.random()]: 'foo'
+}).foo;
+({
+  [Math.random()]: 'foo',
+  foo: 'bar'
+}).foo;
```

Comment:
Could we also test for:

```js
({ foo: fn() }).foo
```

That shouldn't be inlined since `fn()` could cause side effects.

Reply:
It seems that we do the same for arrays : 
`[0, sideEffect()][0]` gets inlined to `0`.

If you still think we should do it I think we can bailout if we recursively find a call expression in the initialiser and maybe share the logic with the array replacer.

Reply:
Ok I didn't know about that. I haven't heard about issues with the array replacer. To me that's safe enough, but a user has no way to know/debug that.

I'm wondering if we shouldn't introduce flags, like GCC's `-O1` to `-Os`.

Reply:
I second both, I think we should warn the user by default and then use multiple optimization level options to silent the warnings and choose if we should assume not side effects.

We could also offer the user to annotate its calls using comments, [à la Closure](https://github.com/google/closure-compiler/wiki/Annotating-JavaScript-for-the-Closure-Compiler). Something like : 

```js

/** @babel:side-effect **/
function foo() {
  sideEffectApi()
}

/** @babel:side-effect:expression:bar() **/
const dontTrim = bar()
```

The is also the [`sideEffects` proposal](https://github.com/webpack/webpack/tree/master/examples/side-effects) from Webpack which may be handy, although I think we're not module aware.

I feel like this needs its own discussion, perhaps I should open an issue?

Reply:
Yeah totally agree, lets create a separate issue and discuss different methods there.